### PR TITLE
PixelDataPool<T>: reduce maximum pooled array size

### PR DESCRIPTION
**Pull request by antonfirsov**
### Prerequisites

- [x] I have written a descriptive pull-request title
- [x] I have verified that there are no overlapping [pull-requests](https://github.com/SixLabors/ImageSharp/pulls) open
- [x] I have verified that I am following matches the existing coding patterns and practice as demonstrated in the repository. These follow strict Stylecop rules :cop:.
- [x] I have provided test coverage for my change (where applicable)

### Description
It might be a longer process to finish #431, however users experiencing `OutOfMemoryException`-s need a stable ImageSharp for yesterday :) In this PR I try to provide a solution for memory issues reported in #225, #409, #325, #224 and #151 by a simple change:
- The maximum pooled buffer size is now limited to 32MB (equivalent to 8 megapixels of raw `Rgba32` data). It's 32-times lower for `Image<Rgba32>` than before. I think most images are going to fit into this size without eating up the memory of the process while *still using pooled memory*. For outliers however, it might be better new-up those arrays. (*You can still run out of usable memory because of LOH fragmentation issues! I suggest to touch [`GCSettings.LargeObjectHeapCompactionMode`](https://blogs.msdn.microsoft.com/mariohewardt/2013/06/26/no-more-memory-fragmentation-on-the-net-large-object-heap/) regularly in environments with strict memory limits!*)

- The same byte limit is used for non-`IPixel` value type arrays. This should probably fix the issues caused by pooled `Block8x8` buffers in JpegDecoder.

We'll keep working on #431 to provide a full-scale solution. @rytmis already did the majority of the work, big thanks for that! (However we still need to fix the array conversion issues, provide proper test coverage, polish the API further before merging it.)

/cc @denisivan0v @xakep139 @vpenades @JulianRooze

### CODE REVIEW

**Comment by denisivan0v** [comment]
Many thanks @antonfirsov, but in my cases it's still a big amount of allocated memory since in [`ArrayPool<T>.Create(MaxArrayLength, 50)`](https://github.com/SixLabors/ImageSharp/pull/436/files#diff-de77e1fc781805efd2c8375e1356c6d4R30) `maxArraysPerBucket = 50` and we end up with 32Mb * 50 * number of buckets of pooled memory.

I'm looking forward to #431 and now rely on these changes to do my tests.
It's also can be very useful to have an API to check image metadata (see https://github.com/SixLabors/ImageSharp/pull/292#issuecomment-357602218) without decoding it and make these allocations.

**Comment by rytmis** [comment]
I've read the ArrayPool API docs and the DefaultArrayPool implementation. Those parameters mean that you've got up to 50 32MB arrays, up to 50 16MB arrays and so on -- the first argument being the max size _within a bucket_ and the second being the maximum number of "similarly-sized" arrays within a single bucket of the pool. The size selection of DefaultArrayPool for a bucket works in powers of two, and you'll always get the size that's the smallest available but at least as large as you asked for.

Out of curiosity, I ran the bucket count and bucket size algorithms for a DefaultArrayPool with those parameters, and the absolute maximum retained size would be just over 3 GB if every single array was in use. Of course I know that it's extremely unlikely that any allocation pattern will hit very close to that mark, but even a 50% fill factor will swallow a large portion of the memory space in 32 bit or in a low-memory container.

Add to this the fact that you've got one ArrayPool for every TPixelData you use within the app: for instance, if all you do is construct Rgba32 images from Jpegs, you'll have one ArrayPool of Block8x8's and another of Rgba32's. -- this is the exact case I had when I came along. This means that the max retained size for the pools is north of 6GB.

I'd suggest that instead of dropping the max size of an individual array, you'd drop the max bucket count instead -- or, possibly, both. If you try to allocate beyond what the pool is capable of holding, it will just return a regular heap-allocated array instead.

(Caveat: I'm currently undergoing a case of the Man Flu, so there's a possibility that I've messed up my math at some point. If you wish to double-check, the sources are at https://github.com/dotnet/corefx/blob/master/src/System.Buffers/src/System/Buffers )

**Comment by antonfirsov** [comment]
@rytmis good experiment and useful calculations! Gonna reduce `maxArraysPerBucket` then, but I think, in usual cases it is very unlikely to get even close to those theoretical limits, specially with the largest 32MB buckets.

I'm also unsure about my math, but if your service is kinda stateless (so buffers are always returned to the pool when a request is finished), you need to stress your application in a way so it reaches a maximum peek of ~10 **concurrent** `~4000x2000` sized image uploads to approach the `50%` `ArrayPool` fill factor.

For me it feels that if this is a realistic scenario in your environment then:
- It is very likely that your gonna experience memory issues even without `ArrayPool` because of [LOH fragmentation](https://blogs.msdn.microsoft.com/maoni/2016/05/31/large-object-heap-uncovered-from-an-old-msdn-article/): The .NET GC [has not been designed](https://github.com/dotnet/corefx/issues/5921#issuecomment-181970876) to follow such a stress.
- For image processing applications, you should really consider using a less limited container, or use load balancing techniques to reduce the load of your individual containers, regardless of the image processing framework you use.

But instead of guessing, I really want to model this by defining proper load testing benchmarks. It's not easy for me to make them realistic, so any input and review would be really appreciated!

**Comment by rytmis** [comment]
My scenario is a batch processor that will download and convert images according to a number of different presets. After each resize, the resized image is stored on disk and the image itself is disposed, so that _should_ mean having _at most_ two images in memory at a time. No concurrent processing.

After running a batch where there were lots of sequential resizes happening, my memory usage peaked somewhere around 1,5 GB and the app began to slow down (I'm guessing GC and disk thrashing due to page file usage).

Note that this is a scenario that has no such issues with System.Drawing. I'm not talking hypotheticals here: what I did was take an existing batch app that targeted the desktop framework and ported it to run on CoreFX, replacing System.Drawing with ImageSharp. The app has been running for over two years without any memory-related issues, but the ported version choked almost immediately.

The main problem the current pool approach is that if you happen to load a single image that is, say, 90 MB in size, you'll _permanently_ retain 128 MB of memory in that pool bracket, even when it has been released back to the pool. If you then load a 130 MB image, you'll permanently retain 256 MB. So going through a series of images that get ever larger, you'll eventually exhaust the heap space even when you'd only need enough capacity for the one you are currently working with. 

In other words, if you ever, at any point, allocate _and then release_ one array from each bucket up until you reach the 130 MB mark, you will have retained at least half a gigabyte until the app terminates. Multiply that by the number of pixel formats, and... you get the picture (pun not intended).

(These calculations assume byte arrays -- I'm not sure, but for TPixel arrays this might change the constants by a little bit. In any case, it's still the same order of magnitude).

**Comment by vpenades** [comment]
@rytmis That's pretty much the same scenario I have. I still believe that using pools for caching lots and lots of very small, similar sized arrays, like handling server messages with lots of small POCOs, it's fine....

But images are too large, if the pool's memory cannot be completely disposed by the GC, or at will, there's many scenarios in which you can end up with your application running with less than half of available memory.

**Comment by antonfirsov** [comment]
@rytmis I think you really need outliers of size `>128MB` to reach that 1.5 GB peek. Now that the maximum pool size is reduced significantly (while still covering typical image sizes), I expect your application's memory usage to be stabilized at much lower levels.

Whether `ArrayPool` is good for imaging or not:
I'm on the strong opinion, that there *always exists an `L > 0` upper limit* for the pool size which leads to better throughput with pooling enabled than without it. In practice, the memory used by the pool should stop growing after a certain number of requests, way beyond it's theoretical limit:

image

I'm quite sure I will be able proove this later with benchmarks.

This limit might be lower for certain applications + scenarios, pretty high for others, but *never zero*, if you optimize for throughput. It's really unfortunate, that bugs we made in ImageSharp (+lack of possibility to customize stuff) led to a debate on ArrayPools with those strong counter-opinions :disappointed: If you are using ASP.NET MVC, you [already have ArrayPools in your application](https://github.com/aspnet/Mvc/search?utf8=%E2%9C%93&q=arraypool&type=)!

**Comment by vpenades** [comment]
@antonfirsov I'm not against ArrayPools, I agree they provide a performance benefit, I also want speed too!. 😄 

What am I against is statically allocated, non disposable array pools. From your chart, what I want is that if you're no longer performing any sort of image processing, the memory usage to drop back to zero.

**Comment by antonfirsov** [comment]
@vpenades totally understand it. [You'll be able to implent this kind of logict](https://github.com/SixLabors/ImageSharp/issues/225#issuecomment-358038796) after merging #431.

**Comment by rytmis** [comment]
Apologies if I sounded like I don't like ArrayPools -- that was never my intention. 🙂 I was just trying to point out potential issues in this particular mitigation strategy, that's all.

**Comment by antonfirsov** [comment]
@rytmis pushed a change based on your observations. Gonna adapt this logic in #431.

**Comment by antonfirsov** [comment]
I'm merging this now. If anyone is about to try our next nightlies because of this change: your feedback is really appreciated.

**Comment by GeorgePlotnikov** [comment]
@antonfirsov below the result

image
the process was finished with 

```
An unhandled exception of type 'SixLabors.ImageSharp.Formats.Jpeg.GolangPort.Components.Decoder.EOFException' occurred in System.Private.CoreLib.dll
Reached end of stream before proceeding EOI marker!
```
